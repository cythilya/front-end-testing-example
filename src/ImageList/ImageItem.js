const ImageItem = ({ title, image }) => {
  return (
    <div data-test-id="image-item">
      <img alt={title} data-test-id="image-item-src" src={image} />
      <p data-test-id="image-item-title">{title}</p>
    </div>
  );
};

export default ImageItem;
