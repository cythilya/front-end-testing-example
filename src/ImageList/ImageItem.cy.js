import ImageItem from './ImageItem';

const MOCK_IMAGE = {
    id: '1',
    title: '打造高速網站，從網站指標開始！全方位提升使用者體驗與流量的關鍵',
    images: ['https://bit.ly/4228IT0'],
  };

describe('ImageItem', () => {
  it('should show image title correctly', () => {
    cy.mount(<ImageItem title={MOCK_IMAGE.title} image={MOCK_IMAGE.images[0]} />);

    cy.get('[data-test-id="image-item-title"]').should('have.text',
      '打造高速網站，從網站指標開始！全方位提升使用者體驗與流量的關鍵'
    );
  });

  it('should show image url correctly', () => {
    cy.mount(<ImageItem title={MOCK_IMAGE.title} image={MOCK_IMAGE.images[0]} />);

    cy.get('[data-test-id="image-item-src"]').should('have.attr', 'src',
      'https://bit.ly/4228IT0'
    );
  });
});
