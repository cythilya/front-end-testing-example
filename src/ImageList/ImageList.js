import { useEffect, useState } from 'react';
import ImageItem from './ImageItem';

const ImageList = () => {
  const [images, setImages] = useState([]);

  useEffect(() => {
    async function fetchData() {
      const response = await fetch('https://dummyjson.com/products');

      const data = await response.json();
      setImages(data.products);
    }
    fetchData();
  }, []);

  return (
    <>
      {!images.length && (
        <div data-test-id="no-data-prompt">No data to display.</div>
      )}
      {images.map(({ id, title, images }) => (
        <ImageItem key={id} title={title} image={images[0]} />
      ))}
    </>
  );
};

export default ImageList;
