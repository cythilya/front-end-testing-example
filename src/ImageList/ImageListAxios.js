import { useEffect, useState } from 'react';
import axios from 'axios';
import ImageItem from './ImageItem';

const ImageList = () => {
  const [images, setImages] = useState([]);

  useEffect(() => {
    async function fetchData() {
      try {
        const response = await axios.get('https://dummyjson.com/products');
        setImages(response.data.products);
      } catch (error) {
        throw error;
      }
    }
    fetchData();
  }, []);

  const renderImages = () => {
    return images.map(({ id, title, images }) => {
      return <ImageItem key={id} title={title} image={images[0]} />;
    });
  };

  return (
    <>
      {images.length !== 0 ? (
        renderImages()
      ) : (
        <div data-test-id="no-data-prompt">No data to display.</div>
      )}
    </>
  );
};

export default ImageList;
