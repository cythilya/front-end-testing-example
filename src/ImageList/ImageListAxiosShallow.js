import { useState } from 'react';
import ImageItem from './ImageItem';

const ImageList = (data) => {
  const [images, setImages] = useState(data.products);

  const renderImages = () => {
    return images.map(({ id, title, images }) => (
      <ImageItem key={id} title={title} image={images[0]} />
    ));
  };

  const renderNoDataPrompt = () => {
    return <div data-test-id="no-data-prompt">No data to display.</div>;
  };

  return <>{!!images.length ? renderImages() : renderNoDataPrompt()}</>;
};

export default ImageList;
