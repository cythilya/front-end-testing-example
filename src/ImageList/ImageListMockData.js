import { useEffect, useState } from 'react';
import axios from 'axios';
import ImageItem from './ImageItem';

const ImageList = ({ data }) => {
  const renderImages = () => {
    return data.products.map(({ id, title, images }) => (
      <ImageItem key={id} title={title} image={images[0]} />
    ));
  };

  const renderNoDataPrompt = () => {
    return <div data-test-id="no-data-prompt">No data to display.</div>;
  };

  return <>{!!data.products.length ? renderImages() : renderNoDataPrompt()}</>;
};

export default ImageList;
