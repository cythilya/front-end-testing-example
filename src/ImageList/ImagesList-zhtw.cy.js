import ImageItem from './ImageItem';

describe('ImageList', () => {
  describe('ImageList', () => {
    it('should render no data prompt when no images', () => {
      // 攔截 API https://dummyjson.com/products 的回應，使用 fixtures 資料夾中的 `empty.json`
      // 渲染 `<ImageList />` 元件
      // 斷言沒有資料時會出現 "no-data-prompt" 提示
    });

    it('should render correct image items when have data', () => {
      // 攔截 API https://dummyjson.com/products 的回應，使用 fixtures 資料夾中的 `empty.json`
      // 渲染 `<ImageList />` 元件
      // 斷言 image items length is 2
    });
  });
});
