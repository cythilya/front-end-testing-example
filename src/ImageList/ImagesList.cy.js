import ImageList from './ImageList';

describe('ImageList', () => {
  it('should render no data prompt when no images', () => {
    cy.intercept('GET', 'https://dummyjson.com/products', {
      statusCode: 200,
      body: { products: [] },
    }).as('apiRequest');

    cy.mount(<ImageList />);

    cy.get('[data-test-id="no-data-prompt"]').should('exist');
  });

  it('should render correct image items when have data', () => {
    cy.intercept('GET', 'https://dummyjson.com/products', {
      statusCode: 200,
      body: {
        products: [
          {
            id: '1',
            title: '打造高速網站，從網站指標開始！全方位提升使用者體驗與流量的關鍵',
            images: ['https://bit.ly/4228IT0'],
          },
          {
            id: '2',
            title: '島波海道。單車跳島之旅',
            images: ['https://bit.ly/4b5nLPO'],
          },
        ],
      },
    }).as('apiRequest');

    cy.mount(<ImageList />);

    cy.get('[data-test-id="image-item-title"]').should('have.length', 2);
    cy.contains('打造高速網站');
    cy.contains('島波海道');
  });
});
