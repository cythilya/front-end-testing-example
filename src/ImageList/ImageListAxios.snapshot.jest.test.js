import React from 'react';
import axios from 'axios';
import { act } from 'react-test-renderer';
import { render } from '@testing-library/react';
import ImageList from './ImageListAxios';

jest.mock('axios');

describe('ImageList', () => {
  it('should render no data prompt when no images', async () => {
    const mockedData = { products: [] };
    axios.get.mockResolvedValue({ data: mockedData });

    const wrapper = await act(async () => render(<ImageList />));

    expect(wrapper).toMatchSnapshot();
  });

  it('should render correct image items when have data', async () => {
    const mockedData = {
      products: [
        {
          id: '1',
          title:
            '打造高速網站，從網站指標開始！全方位提升使用者體驗與流量的關鍵',
          images: ['https://bit.ly/4228IT0'],
        },
        {
          id: '2',
          title: '島波海道。單車跳島之旅',
          images: ['https://bit.ly/4b5nLPO'],
        },
      ],
    };
    axios.get.mockResolvedValue({ data: mockedData});

    const wrapper = await act(async () => render(<ImageList />));

    expect(wrapper).toMatchSnapshot();
  });
});
