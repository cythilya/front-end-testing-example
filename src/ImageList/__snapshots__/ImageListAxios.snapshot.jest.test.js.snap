// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageList should render correct image items when have data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-test-id="image-item"
      >
        <img
          alt="打造高速網站，從網站指標開始！全方位提升使用者體驗與流量的關鍵"
          data-test-id="image-item-src"
          src="https://bit.ly/4228IT0"
        />
        <p
          data-test-id="image-item-title"
        >
          打造高速網站，從網站指標開始！全方位提升使用者體驗與流量的關鍵
        </p>
      </div>
      <div
        data-test-id="image-item"
      >
        <img
          alt="島波海道。單車跳島之旅"
          data-test-id="image-item-src"
          src="https://bit.ly/4b5nLPO"
        />
        <p
          data-test-id="image-item-title"
        >
          島波海道。單車跳島之旅
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-test-id="image-item"
    >
      <img
        alt="打造高速網站，從網站指標開始！全方位提升使用者體驗與流量的關鍵"
        data-test-id="image-item-src"
        src="https://bit.ly/4228IT0"
      />
      <p
        data-test-id="image-item-title"
      >
        打造高速網站，從網站指標開始！全方位提升使用者體驗與流量的關鍵
      </p>
    </div>
    <div
      data-test-id="image-item"
    >
      <img
        alt="島波海道。單車跳島之旅"
        data-test-id="image-item-src"
        src="https://bit.ly/4b5nLPO"
      />
      <p
        data-test-id="image-item-title"
      >
        島波海道。單車跳島之旅
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ImageList should render no data prompt when no images 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-test-id="no-data-prompt"
      >
        No data to display.
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-test-id="no-data-prompt"
    >
      No data to display.
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
